Make AuthorizationProxyFactory.proxy generic

big-cir opened the pull request:
Overview
---
This pull request updates the AuthorizationProxyFactory interface to make its proxy method generic, ensuring that it returns the same type as the input object. 

The following changes are included:
- Refactored AuthorizationProxyFactory.proxy to use generics (<T> T proxy(T object)).
- Removed redundant casts from production and test code that are no longer necessary due to the type-safe proxy method.

<br>

Related
---
ISSUE: #16706 

<br>

Note
---
I’m aware that the issue is currently labeled with status: waiting-for-triage, which indicates that the final design direction may not be settled yet. This PR reflects my current understanding of the issue and is intended as a starting point for further discussion.

If this approach does not align with the intended direction, please feel free to close the PR or provide feedback. I am happy to revise or take a different direction as needed. Thank you!

jzheaux commented on core/src/test/java/org/springframework/security/authorization/AuthorizationAdvisorProxyFactoryTests.java:
```diff
@@ -335,7 +335,7 @@ public void setTargetVisitorThenUses() {
 	public void setTargetVisitorIgnoreValueTypesThenIgnores() {
 		AuthorizationAdvisorProxyFactory factory = AuthorizationAdvisorProxyFactory.withDefaults();
-		assertThatExceptionOfType(ClassCastException.class).isThrownBy(() -> ((Integer) factory.proxy(35)).intValue());
+		assertThatExceptionOfType(ClassCastException.class).isThrownBy(() -> (factory.proxy(35)).intValue());
```
It looks like the parentheses are no longer needed around `factory.proxy(35)`. Can you remove them please?

jzheaux reviewed (changes requested):
Thanks, @big-cir! I've left some feedback inline.

big-cir commented:
> Thanks, @big-cir! I've left some feedback inline.

@jzheaux Thank you for your valuable feedback.
I've updated the code according to your comments.
Please review the changes and let me know if any further modifications are needed. :)

jzheaux commented:
Thanks, @big-cir! This looks great. In preparation for merging, will you please squash your commits and then rewrite it to include the ticket number like so:

```
Make AuthorizationProxyFactory#proxy Generic

Closes gh-16706
```

big-cir commented:
> Thanks, @big-cir! This looks great. In preparation for merging, will you please squash your commits and then rewrite it to include the ticket number like so:
> 
> ```
> Make AuthorizationProxyFactory#proxy Generic
> 
> Closes gh-16706
> ```

I've squashed the commits and updated the commit message as requested.
Thank you for the review! :)
